pkg/sources/udsource/user_defined_source.go

```diff
@@ -130,15 +121,28 @@ func (u *userDefinedSource) GetName() string {
 	return u.vertexName
 }
 
-// GetPartitionIdx returns the partition number for the user-defined source.
-// Source is like a buffer with only one partition. So, we always return 0
-func (u *userDefinedSource) GetPartitionIdx() int32 {
-	return 0
+// Partitions returns the partitions of the user-defined source
+func (u *userDefinedSource) Partitions() []int32 {
+	partitions := make([]int32, 0, len(u.partitions))
+	for partition := range u.partitions {
+		partitions = append(partitions, partition)
+	}
+	return partitions
 }
 
-// Read reads the messages from the user-defined source
+// Read reads the messages from the user-defined source, tracks the partitions from which the messages are read
+// tracked partitions are used to determine the partitions to which the watermarks should be published
 func (u *userDefinedSource) Read(ctx context.Context, count int64) ([]*isb.ReadMessage, error) {
-	return u.sourceApplier.ApplyReadFn(ctx, count, u.readTimeout)
+	readMessages, err := u.sourceApplier.ApplyReadFn(ctx, count, u.readTimeout)
+	if err != nil {
+		return nil, err
+	}
+	for _, msg := range readMessages {
+		if _, ok := u.partitions[msg.ReadOffset.PartitionIdx()]; !ok {
+			u.partitions[msg.ReadOffset.PartitionIdx()] = struct{}{}
+		}
+	}
```

Comment:
this is not right, we only append by never remove. we will have to ask users to add `Partitions()` to ud-source interface. 

@KeranYang WDYT?

Reply:
#1405

Reply:
I agree this is not right. With this change we now have two sources of truth on how many partitions a source has. One is the read batch and the other is the source Partitions() interface.

The removing part can be tricky because if we remove a certain partition, we also need to remove the corresponding watermark KV store etc. We can look into details as part of issue 1405 that you created.